### PR TITLE
feat: add support for response compression

### PR DESCRIPTION
**Pull request by Rutik7066**
/claim #602 

We have added middleware to compress the response. 

## TODO

- [x] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: router/core/header_rule_engine_test.go
Comment (StarpTech) [inline router/core/header_rule_engine_test.go:327]: Why does the branch contains changes from your work on the rename header rule?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/header_rule_engine_test.go
Comment (Rutik7066) [reply router/core/header_rule_engine_test.go:327]: Actually I worked on same code base and then published the changes to the new branch and created the draft pr.

**Comment by Rutik7066** [comment]
Hi @StarpTech If you like I can claim #674 in this repo.

**Comment by Rutik7066** [comment]
@StarpTech  with the current implementation We would have to add '''Content-Type''' in response header I tried to add middleware for it but failed to do it. Could you please check the latest changes?

File: router/core/router.go
Comment (StarpTech) [inline router/core/router.go:931]: Please ensure that also GraphQL specific content types are added to the list of supported compression content types https://github.com/go-chi/chi/blob/ef31c0bff3f8061e6fba7085691e4970a3d1b7da/middleware/compress.go#L17

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Comment by StarpTech** [comment]
Please rebase.

File: router/go.sum
Comment (Rutik7066) [reply router/go.sum:390]: Yes, we are using https://github.com/andybalholm/brotli which pure go implementation of https://github.com/google/brotli. I have removed gopkg.in/kothar/brotli-go.v0  it.

**Comment by Rutik7066** [comment]
@StarpTech Thank you for review. all the requested changes  is completed.
Screenshot 2024-05-18 154647

**Comment by Rutik7066** [comment]
@StarpTech tests are failing for Authentication. I don't know why this is happening. I did not make any changes to the Authentication module. can please check it.

**Comment by Rutik7066** [comment]
@StarpTech working perfectly now.

**Comment by StarpTech** [comment]
Sounds great. I think we can merge it this week. I'll do a last test tomorrow.

**Comment by Rutik7066** [comment]
@StarpTech did you tested the code?

File: router-tests/compression_test.go
Comment (StarpTech) [inline router-tests/compression_test.go:16]: The function has to start with `Test`

File: router-tests/integration_test.go
Comment (StarpTech) [inline router-tests/integration_test.go:1182]: Please put the tests into a separate file e.g `request_compression_test.go`

File: router/core/router.go
Comment (StarpTech) [inline router/core/router.go:1103]: This should be defined in the corresponding package.

File: router/internal/graphiql/playgroundhandler.go
Comment (StarpTech) [inline router/internal/graphiql/playgroundhandler.go:24]: Revert

File: router-tests/compression_test.go
Comment (StarpTech) [inline router-tests/compression_test.go:18]: Unit tests should be in the router package. I recommend to create a compression package where you can define the code and variables.

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Comment by StarpTech** [comment]
Almost done. I found a couple of other things.

File: router/core/router.go
Comment (Rutik7066) [reply router/core/router.go:1103]: Package defaults does not contain graphql types and if we pass contain types then package will only use that types instead of default. In our case we need both so we have pass all types including graphql types.

**Comment by Rutik7066** [comment]
Working on it. It will be done today only.

File: router/core/router.go
Comment (StarpTech) [reply router/core/router.go:1103]: My point is that the router module should not host those variables. It would be good, if you put those variables into a dedicated package.

Comment (Rutik7066) [reply router/core/router.go:1103]: Sure no problem

**Comment by Rutik7066** [comment]
@StarpTech Changes applied. Please check.

File: router/go.mod
Comment (StarpTech) [inline router/go.mod:26]: Please revert these changes.

File: router/core/request_compression_test.go
Comment (StarpTech) [inline router/core/request_compression_test.go:27]: I think it would makes sense, if we would validate all possible content types `"deflate", "gzip", "br"` before we add the compression middleware

File: router/core/compressible_content_types.go
Comment (StarpTech) [inline router/core/compressible_content_types.go:1]: This is the same package. Then you also define it [here](https://github.com/wundergraph/cosmo/blob/f4cbc28a3a51779eeaaa5108f44e56413caaf005/router/core/router.go#L72) instead to create a new file.

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Comment by Rutik7066** [comment]
@StarpTech  changes applied. Github action Router CI is falling don't know why. It is not related to my code.  
After updating branch.

**Comment by StarpTech** [comment]
@Rutik7066 https://github.com/wundergraph/cosmo/pull/723#discussion_r1613153286

File: router/core/request_compression_test.go
Comment (Rutik7066) [reply router/core/request_compression_test.go]: As you instructed, I have added code to validate all content types.

Comment (Rutik7066) [reply router/core/request_compression_test.go]: Have I misunderstood?

File: router/core/request_compression_test.go
Comment (StarpTech) [inline router/core/request_compression_test.go:63]: @Rutik7066 here why is "br" not part of the list?

Comment (Rutik7066) [reply router/core/request_compression_test.go:63]: deflate and gzip are supported by go-chi router. so we have to add brotli compression as custom middleware.

Comment (StarpTech) [reply router/core/request_compression_test.go:63]: How is it related to the middleware that validates what content-encoding is allowed?

File: router/core/request_compression_test.go
Comment (StarpTech) [inline router/core/request_compression_test.go:31]: What is the value of unit testing if dependencies are working? Please put the most important tests into integration tests.

File: router-tests/request_compression_test.go
Comment (StarpTech) [inline router-tests/request_compression_test.go:1]: @Rutik7066 I'm slightly confused. This tests response compression.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/request_compression_test.go
Comment (Rutik7066) [reply router/core/request_compression_test.go:31]: I also feel that unit test for this is unnecessary. Only integration test is necessary. I will remove the unit test.

**Comment by Aenimus** [comment]
@Rutik7066 can you run `make sync-go-workspace` from root please?

**Comment by Rutik7066** [comment]
> make sync-go-workspace

@Aenimus  Done.

File: router/core/request_compression_test.go
Comment (Rutik7066) [reply router/core/request_compression_test.go]: Yes your are right . it should be before compressor middleware and br should be added in the list

**Comment by Rutik7066** [comment]
@StarpTech  we can merge this now. all changes are applied. please check.

**Comment by StarpTech** [comment]
@Rutik7066 I can't merge the PR. The PR title and tests files says "request compression" ([example](https://echo.labstack.com/docs/middleware/decompress)) but you addressed response compression. The added middleware `AllowContentEncoding` validates if a request in a specific encoding can be decoded but you did not add support for it. 

I should remind you that the bounty is designed to be a solo challenge. The requirements are pretty clear.

In order to merge the PR, please use the wording "response compression" everywhere in test files, PR etc.. Additionally, remove the `AllowContentEncoding` middleware which does nothing. After that we can merge only "response compression"

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Comment by Rutik7066** [comment]
Ok I will update.

File: router-tests/request_compression_test.go
Comment (Rutik7066) [reply router-tests/request_compression_test.go]: added zstd for testing unsupported encoding and identity for no encoding.

**Comment by Rutik7066** [comment]
Hi @StarpTech  all changes applied.

**Comment by Rutik7066** [comment]
@StarpTech I hope everything is perfect now to merge.

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**
